allow input of readonly arrays

EdJoPaTo opened the pull request:
In my opinion this method (alwaysArray) is meant for function arguments to be more versatile.
But the internal code of the user wants to always have an array.

Using a readonly array as input allows for even more possible inputs. (writeable arrays can also be used in readonly arrays but not the other way around)
This has the drawback of only a readonly output array.
Internally the array from input shouldn't be changed anyway so this shouldn't be a problem.

BREAKING CHANGE: returns only a readonly array
This will result in a version 2.0.0 but this should not impact any good code as the input of a method shouldn't be changed anyway.

EdJoPaTo committed:
chore(npm): don't create package-lock

EdJoPaTo committed:
chore(npm): build before packing / publishing

EdJoPaTo committed:
feat: allow readonly input array

This method is meant for function arguments to be more versatile.
But the internal code wants to always have an array.

Using a readonly array as input allows for even more possible inputs.
This has the drawback of only a readonly output array.
Internally the array from input shouldnt be change anyway so this
shouldn't be a problem.

BREAKING CHANGE: returns only a readonly array

EdJoPaTo committed:
feat: export type of input

krmax44 commented on src/index.ts:
```diff
@@ -1,4 +1,6 @@
-export default function alwaysArray<T>(input: T | T[]): T[] {
+export type SingleOrArray<T> = T | readonly T[];
+
+export default function alwaysArray<T>(input: SingleOrArray<T>): readonly T[] {
```
What about removing the visibility from the return value?

```suggestion
export default function alwaysArray<T>(input: SingleOrArray<T>): T[] {
```

From my testing, readonly attributes can still be set to non-readonly values:

```ts
interface TestObject {
	readonly value: number[];
}

const a: TestObject = {
	value: [1, 2, 3, 4, 5]
};
const b = alwaysArray(a.value); // [1, 2, 3, 4, 5]
const c: TestObject = {
	value: b // seems to be okay
};
```

EdJoPaTo replied:
I think you mixed up the `readonly` keyword here.
The interface in your example has a readonly member with a normal array. It can be set only once but the array itself is writeable.

```ts
interface TestObjectA {
	readonly value: number[];
}

interface TestObjectB {
	value: readonly number[];
}

const a: TestObjectA = {value: [1, 2, 3]};
const b: TestObjectB = {value: [1, 2, 3]};

a.value = []; // fails as the member is readonly
b.value = []; // works as the member is not readonly

a.value.push(5); // works as the variable itself is writeable
b.value.push(5); // is denied as the variable itself is readonly
```

Maybe this example can explain why I requested this change:

```ts
// Method to get some info of an array. This method will not modify the array, so it can be readonly (and should be to let the compiler warn me about changing it.)
function onlyRead(values: readonly number[]): number {
	const foo = someTotallyUnrelatedMethodInATotallyDifferentContext(values)
	const bar = other(values) // does not work
	return foo * bar
}

function someTotallyUnrelatedMethodInATotallyDifferentContext(values: number | readonly number[]) {
	const arr = alwaysArray(values)
	return arr.length
}

function other(values: number | number[]) {
	const arr = alwaysArray(values)
	return arr.length
}
```

Currently the method returns the array directly when the argument is an array: `Array.isArray(input) === true` → `return input`. This returns the input with its type without changing it (readonly array). The return type of the method has to be the same.
The return type could be opened up to be writeable too (`Array.isArray(input) ? […input] : [input]`) but this would introduce an additional content "copy" (as far as JavaScript actually copies and not just references stuff).

As this is used on arguments for methods anyway I don't think it should be writeable anyway so I would keep It readonly.

I hope I could explain my point in a better way now :)

krmax44 replied:
Ah, I see. I'd prefer returning a writeable array just to avoid confusion. Maybe we can combine this with #11, where I implemented spreading of iterables, so adding that for arrays as well was trivial: https://github.com/krmax44/always-array/blob/3675f939277f99d721b27d4848ea3c0b4aaac6e6/src/index.ts

Does this solve the issue for you?

krmax44 commented on package.json:
```diff
@@ -13,6 +13,7 @@
     "test": "mocha -r ts-node/register 'src/**/*.test.ts' && yarn lint",
     "lint": "xo",
     "build": "tsc -p ./tsconfig.json && uglifyjs build/index.js -o build/index.js -m -c",
+    "prepare": "npm run build",
```
```suggestion
    "prepare": "yarn build",
```

EdJoPaTo replied:
When scripts do stuff it should be included in the devDependencies. `yarn` is currently not included there. Shall I change that?

krmax44 replied:
We should probably include that besides Node.js, Yarn needs to be installed as well. Installing Yarn via npm isn't [recommended](https://stackoverflow.com/a/43377805) and it might conflict with the system-wide install vs. the one in `node_modules`.

Or we just keep `npm run build`, I don't really mind that much.

EdJoPaTo replied:
I prefer to have repositories self contained. They shouldn't need something specific installed globally in my opinion. Especially as dependency versions might differ in different repositories.

EdJoPaTo committed:
chore: built tests but dont pack them

they were not built as they were excluded

EdJoPaTo committed:
chore(typescript): enable strict compiling

Libraries published on npm should be as strict as possible to
prevent possible errors later on.

EdJoPaTo committed:
fix: always return writeable array

krmax44 commented:
included in #11
